electrum/address_synchronizer.py

```diff
@@ -338,35 +339,41 @@ def get_depending_transactions(self, tx_hash):
                 children |= self.get_depending_transactions(other_hash)
             return children
 
-    def receive_tx_callback(self, tx_hash, tx, tx_height):
-        self.add_unverified_tx(tx_hash, tx_height)
+    def receive_tx_callback(self, tx_hash, tx, tx_height, tx_type=TxType.NONVAULT):
+        self.add_unverified_tx(tx_hash, tx_height, tx_type.name)
         self.add_transaction(tx_hash, tx, allow_unrelated=True)
 
     def receive_history_callback(self, addr: str, hist, tx_fees: Dict[str, int]):
         with self.lock:
             old_hist = self.get_address_history(addr)
-            for tx_hash, height, *__ in old_hist:
-                if (tx_hash, height) not in hist:
+            for tx_hash, height, tx_type in old_hist:
+                if (tx_hash, height, tx_type) not in hist:
                     # make tx local
                     self.unverified_tx.pop(tx_hash, None)
                     self.db.remove_verified_tx(tx_hash)
                     if self.verifier:
                         self.verifier.remove_spv_proof_for_tx(tx_hash)
             self.db.set_addr_history(addr, hist)
 
-        for tx_hash, tx_height, *__ in hist:
+        for tx_hash, tx_height, tx_type in hist:
             # add it in case it was previously unconfirmed
-            self.add_unverified_tx(tx_hash, tx_height)
+            self.add_unverified_tx(tx_hash, tx_height, tx_type)
             # if addr is new, we have to recompute txi and txo
             tx = self.db.get_transaction(tx_hash)
             if tx is None:
                 continue
+            self._mutate_transaction_type(current_tx=tx, incoming_type=TxType.from_str(tx_type))
             self.add_transaction(tx_hash, tx, allow_unrelated=True)
 
         # Store fees
         for tx_hash, fee_sat in tx_fees.items():
             self.db.add_tx_fee_from_server(tx_hash, fee_sat)
 
+    def _mutate_transaction_type(self, current_tx: 'ThreeKeysTransaction', incoming_type: TxType):
```

Comment:
will it always be ThreeKeysTransaction? remember that from now on this logic happens in all wallet types

Reply:
every fetched transaction via `self.network.get_transaction(tx_hash)` is being casted into `ThreeKeysTransaction`